# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: docshell/base/WindowContext.h
# Commit: 52ac63af39a4
# Full Hash: 52ac63af39a4de5a2e8ea8d44071a3006f6a1b8d
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2020-04-25 09:49:23
# Regressor Bug: 1580565
# File Overlap Count: 3
# Description:
#   Bug 1580565 - Part 4: Use WindowContext to manage BrowsingContext cached status, r=farre
#   
#   The existing infrastructure which stored cached BrowsingContexts on the
#   BrowsingContextGroup was added before WindowContexts were added, and can cause
#   racing issues with partially discarded trees during process switches.
# ==============================================================================

diff -r 775fcd064844 -r 52ac63af39a4 docshell/base/WindowContext.h
--- a/docshell/base/WindowContext.h	Fri Apr 24 18:33:01 2020 +0000
+++ b/docshell/base/WindowContext.h	Fri Apr 24 18:33:04 2020 +0000
@@ -7,6 +7,7 @@
 #ifndef mozilla_dom_WindowContext_h
 #define mozilla_dom_WindowContext_h
 
+#include "mozilla/Span.h"
 #include "mozilla/dom/MaybeDiscarded.h"
 #include "mozilla/dom/SyncedContext.h"
 #include "mozilla/net/NeckoChannelParams.h"
@@ -44,6 +45,10 @@
   uint64_t OuterWindowId() const { return GetOuterWindowId(); }
   bool IsDiscarded() const { return mIsDiscarded; }
 
+  bool IsCached() const;
+
+  Span<RefPtr<BrowsingContext>> Children() { return mChildren; }
+
   // Cast this object to it's parent-process canonical form.
   WindowGlobalParent* Canonical();
 
@@ -58,6 +63,12 @@
     uint64_t mBrowsingContextId;
 
     FieldTuple mFields;
+
+    bool operator==(const IPCInitializer& aOther) const {
+      return mInnerWindowId == aOther.mInnerWindowId &&
+             mBrowsingContextId == aOther.mBrowsingContextId &&
+             mFields == aOther.mFields;
+    }
   };
   IPCInitializer GetIPCInitializer() {
     return {mInnerWindowId, mBrowsingContext->Id(), mFields.Fields()};
@@ -74,6 +85,11 @@
   void Init();
 
  private:
+  friend class BrowsingContext;
+
+  void AppendChildBrowsingContext(BrowsingContext* aBrowsingContext);
+  void RemoveChildBrowsingContext(BrowsingContext* aBrowsingContext);
+
   // Send a given `BaseTransaction` object to the correct remote.
   void SendCommitTransaction(ContentParent* aParent,
                              const BaseTransaction& aTxn, uint64_t aEpoch);
@@ -114,6 +130,12 @@
   uint64_t mInnerWindowId;
   RefPtr<BrowsingContext> mBrowsingContext;
 
+  // --- NEVER CHANGE `mChildren` DIRECTLY! ---
+  // Changes to this list need to be synchronized to the list within our
+  // `mBrowsingContext`, and should only be performed through the
+  // `AppendChildBrowsingContext` and `RemoveChildBrowsingContext` methods.
+  nsTArray<RefPtr<BrowsingContext>> mChildren;
+
   bool mIsDiscarded = false;
 };
 